packages/components/src/DateTimeInput.tsx

```diff
@@ -0,0 +1,159 @@
+  className?: string;
+  onChange?(value?: string): void;
+  defaultValue?: string;
+  onFocus?(): void;
+  onBlur?(): void;
+  'data-testid'?: string;
+};
+
+export function addSeparators(value: string): string {
+  const dateTimeMillis = value.substring(0, 23);
+  const micros = value.substring(23, 26);
+  const nanos = value.substring(26);
+  return [dateTimeMillis, micros, nanos].filter(v => v !== '').join('\u200B');
+}
+
+const removeSeparators = (value: string) => value.replace(/\u200B/g, '');
+
+const EXAMPLES = [addSeparators(DEFAULT_VALUE_STRING)];
+
+// Forward ref causes a false positive for display-name in eslint:
+// https://github.com/yannickcr/eslint-plugin-react/issues/2269
+// eslint-disable-next-line react/display-name
+const DateTimeInput = React.forwardRef<HTMLInputElement, DateTimeInputProps>(
+  (props: DateTimeInputProps, ref) => {
+    const {
+      className = '',
+      onChange = () => false,
+      defaultValue = '',
+      onFocus = () => false,
+      onBlur = () => false,
+      'data-testid': dataTestId,
+    } = props;
+    const [value, setValue] = useState(
+      defaultValue.length > 0 ? addSeparators(defaultValue) : ''
+    );
+    const [selection, setSelection] = useState<SelectionSegment>();
+
+    function getNextNumberSegmentValue(
+      delta: number,
+      segmentValue: string,
+      lowerBound: number,
+      upperBound: number,
+      length: number
+    ) {
+      const modValue = upperBound - lowerBound + 1;
+      const newSegmentValue =
+        ((((parseInt(segmentValue, 10) - delta - lowerBound) % modValue) +
+          modValue) %
+          modValue) +
+        lowerBound;
+      const result = `${newSegmentValue}`.padStart(length, '0');
+      log.debug('getNextNumberSegmentValue', modValue, newSegmentValue, result);
+      return result;
+    }
+
+    function getNextSegmentValue(
+      range: SelectionSegment,
+      delta: number,
+      segmentValue: string
+    ): string {
+      const { selectionStart } = range;
+      if (selectionStart === 0) {
+        return getNextNumberSegmentValue(delta, segmentValue, 1900, 2099, 4);
+      }
+      if (selectionStart === 5) {
+        return getNextNumberSegmentValue(delta, segmentValue, 1, 12, 2);
+      }
+      if (selectionStart === 8) {
+        return getNextNumberSegmentValue(delta, segmentValue, 1, 31, 2);
+      }
+      if (selectionStart === 11) {
+        // Hours input
+        return getNextNumberSegmentValue(delta, segmentValue, 0, 23, 2);
+      }
+      if (selectionStart === 17 || selectionStart === 14) {
+        // Minutes/seconds input
+        return getNextNumberSegmentValue(delta, segmentValue, 0, 59, 2);
+      }
+      if (
+        selectionStart === 20 ||
+        selectionStart === 24 ||
+        selectionStart === 28
+      ) {
+        // Milli, micro, and nanosecond input
+        return getNextNumberSegmentValue(delta, segmentValue, 0, 999, 3);
+      }
+
+      return segmentValue;
+    }
```

Comment:
Pull out of component, and export either from this file or a separate `DateTimeUtils.js` file so this can be shared with `DateTime`